### PR TITLE
optional-chaining

### LINKED ISSUE #1909: ES2020 optional chaining instead of guard
It would be nice if decaffeinate generates optional chaining that is introduced in ES2020, and is supported by the latest browsers and Babel/TypeScript.

### PR DESCRIPTION
**Pull request by aminya**
an attempt to fix #1909 
- Properties optional chaining support
- Fixes cli tests

Note: both properties and function call optional chaining are part of the proposal:
https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining

But this PR only implements properties optional chaining.

### CODE REVIEW

**Comment by eventualbuddha** [comment]
Thanks! This looks like a good start, but needs tests. Some examples:

- `a?.b`
- `a?.[b]`
- `a?.b?.c`
- `a?.b()`
- `a?.()`
- `a.b?.()`
- `a?.b?.()`

**Comment by aminya** [comment]
I don't know how you can pass compiler options to the tests! @eventualbuddha Could you help with that?

 This PR just skips properties conversion, so the tests are trivial.

**Comment by aminya** [comment]
@eventualbuddha I am not familiar with the structure of the tests. I have enabled edit access. Could you help with setting up the test?

**Comment by eventualbuddha** [comment]
@aminya I added some commits that should at least point you in the right direction. You’ll need to finish adding the expected soak test case results when using optional chaining in addition to the actual implementation. Let me know if you have more questions.

**Comment by aminya** [comment]
@eventualbuddha I fixed the CLI tests that were failing for me.

I added the optional chaining test starting from https://github.com/decaffeinate/decaffeinate/pull/1969/commits/e747cf071ad92435d220702edc240408abbe6ba6.

Some of them fail because `AddVariableDeclarationsStage` fails to parse that. Probably you should add https://babeljs.io/docs/en/next/babel-plugin-syntax-optional-chaining.html to your parser.

**Comment by krissrex** [comment]
Do you have any tests that try to convert coffee `someArray?[0]` to js `someArray?.[0]` ?

Ref https://github.com/decaffeinate/decaffeinate/issues/1909#issuecomment-656015308 and https://github.com/decaffeinate/decaffeinate/issues/1909#issuecomment-656033291

**Comment by eventualbuddha** [comment]
This turned out to be not that straightforward. You can't simply leave the CoffeeScript code alone since in some cases the JavaScript equivalent has an extra `.`. Still, that part was pretty easy. The harder part is that CoffeeScript soaked expressions can be used in place that JavaScript optional chaining cannot, such as in an assignee: `a?.b = 1` is valid CoffeeScript, but not JavaScript. To address this I added `isInsideAssignee` to check whether a node should be considered part of an assignee or not, and to bail early if it is. Otherwise you'd end up with a parse error in a later stage.

As for actually landing this, I'd like to make this PR focus only on optional chaining. I removed some of the unrelated commits already, but it's still failing for me locally. Can you remove the other changes you made, @aminya? They'd be welcome as separate PRs.

File: src/stages/main/patchers/SoakedFunctionApplicationPatcher.ts [hunk 1]
```diff
@@ -151,6 +158,13 @@ export default class SoakedFunctionApplicationPatcher extends FunctionApplicatio
     soakContainer.appendDeferredSuffix(')');
   }
 
+  /**
+   * Change a[b]?() to a[b]?.()
+   */
+  // patchDynamicMethodCallOptionalChaining(fn: MemberAccessOpPatcher): void {
+  //   // TODO
+  // }
+
```
Comment (eventualbuddha) [inline src/stages/main/patchers/SoakedFunctionApplicationPatcher.ts:167]: Remove this.

File: src/stages/main/patchers/SoakedFunctionApplicationPatcher.ts [hunk 2]
```diff
@@ -121,6 +121,13 @@ export default class SoakedFunctionApplicationPatcher extends FunctionApplicatio
     soakContainer.appendDeferredSuffix(')');
   }
 
+  /**
+   * Change a.b?() to a.b?.()
+   */
+  // patchMethodCallOptionalChaining(fn: MemberAccessOpPatcher): void {
+  //   // TODO
+  // }
+
```
Comment (eventualbuddha) [inline src/stages/main/patchers/SoakedFunctionApplicationPatcher.ts:130]: Remove this.

File: src/stages/main/patchers/SoakedMemberAccessOpPatcher.ts [hunk 3]
```diff
@@ -9,7 +9,7 @@ const GUARD_HELPER = `function __guard__(value, transform) {
 }`;
 
 export default class SoakedMemberAccessOpPatcher extends MemberAccessOpPatcher {
-  _shouldSkipSoakPatch = false;
+  _shouldSkipSoakPatch = this.options.optionalChaining;
```
Comment (eventualbuddha) [inline src/stages/main/patchers/SoakedMemberAccessOpPatcher.ts:12]: Revert this.

File: test/cli_test.ts [hunk 4]
```diff
@@ -4,19 +4,24 @@ import { existsSync, readFileSync } from 'fs';
 import { copySync } from 'fs-extra';
 
 import stripSharedIndent from '../src/utils/stripSharedIndent';
+import { toUnix } from './utils/toUnix';
```
Comment (eventualbuddha) [inline test/cli_test.ts:7]: These changes belong in another PR.

**Comment by aminya** [comment]
Thanks for the update. I am going to make another pull request that fixed the CLI tests. I am not sure what you have removed at this point. Nothing is apparent from the commits. Fortunately, I have a local backup.

**Comment by eventualbuddha** [comment]
The commits are still here on this PR, or retrieving them locally works too. Thanks for getting this started!
